src/weights.jl

```diff
@@ -5,15 +5,18 @@ abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: Abstrac
     @weights name
 
 Generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
-and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
+and stores the `values` (`V<:RealVector` using 1-based indexing) and `sum` (`S<:Real`).
 """
 macro weights(name)
     return quote
         mutable struct $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
             values::V
             sum::S
         end
-        $(esc(name))(values::AbstractVector{<:Real}) = $(esc(name))(values, sum(values))
+        function $(esc(name))(values::AbstractVector{<:Real})
```

Comment:
An alternative that would be safer would be to define inner constructor:
```
macro weights(name)
    return quote
        mutable struct $(esc(name)){S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
            values::V
            sum::S

            function $(esc(name)){S, T, V}(values::AbstractVector{T}, s::S) where {S<:Real, T<:Real, V<:AbstractVector{T}}
                Base.require_one_based_indexing(values)
                return new(values, s)
            end
        end

        $(esc(name))(values::AbstractVector{<:Real}, s=sum(values)) =
            $(esc(name)){typeof(s), eltype(values), typeof(values)}(values, s)
    end
end
```
but this looses all promotion rules that default inner constructor comes with.

Reply:
Given that we want to require 1-based indexing in all cases, the inner constructor approach sounds safer. How many methods do we need to define to replicate the default inner constructor?

Reply:
I have closed the PR since it seems we want to support non-1-based indexing. I think this issue should be settled first.